### PR TITLE
add miniasync hashmap blog post

### PR DESCRIPTION
**Pull request by kswiecicki**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237)

### CODE REVIEW

**Review by pbalcer — CHANGES_REQUESTED** [review]
I like it, great job.

> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237#-:-N0VbiOyEvWvZQsHhvpj:bhggac0)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[`content/blog/2022/masync-ex-hashmap.md` line 64 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0V_k1t9vnXZVxN_waz:-N0V_k1t9vnXZVxN_wb-:b-rrchmc) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L64)):*
> ```Markdown
> 
> The lookup data `struct hashmap_lookup_data` contains a pointer to hashmap `hm`, `key`, and `state` of the
> searched entry. `hm` should point to an actual hashmap instance. `key` is an entry key used in the hash function.
> ```

You are using `.` as `,`.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 99 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Va6zgCef4SWtLqxb1:-N0Va6zgCef4SWtLqxb2:b39ijaj) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L99)):*
> ```Markdown
> means that it has already completed, and `FUTURE_STATE_RUNNING` signalizes that it's still running.
> 
> The `lookup` future task function:
> ```

I suggest starting this explanation by first describing hash maps and linear probing. In general, just some intro about what the example is and how it's meant to work. And then you map that onto your actual implementation.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 258 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vb3rg6SVuWYO8sulP:-N0Vb3rg6SVuWYO8sulQ:b-po3t9s) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L258)):*
> ```Markdown
> struct hashmap_lookup_lock_entry_data {
> 	FUTURE_CHAIN_ENTRY(struct hashmap_lookup_fut, lookup);
> 	FUTURE_CHAIN_ENTRY(struct hashmap_entry_set_state_fut, set_state);
> ```

It would be useful to mention that CHAIN_ENTRY_LAST can be used here if there's a need to store some more state inside of the data struct.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 441 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0VbX1z7eBc4aNhNiIO:-N0VbX1z7eBc4aNhNiIP:bbk30y) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L441)):*
> ```Markdown
> address pointed by the `future` parameter using the `memset(...)` function. `future` points to the
> future entry that we are currently initializing.
> ```

This ends rather abruptly. Add a summary section with that you talked about and where there's more information.

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237)* status: all files reviewed, 12 unresolved discussions (waiting on @kswiecicki)

___
*[`content/blog/2022/masync-ex-hashmap.md` line 6 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0VilBaCqfp-MrbFSLt:-N0VilBaCqfp-MrbFSLu:b61yo08) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L6)):*
> ```Markdown
> 
> # Blog post creation date
> date: 2022-03-30T15:11:18+02:00
> ```

you can bump this to something more recent when you'll have a final version of this post :wink: 
___
*[`content/blog/2022/masync-ex-hashmap.md` line 46 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vk7683Qq6xT653drP:-N0Vk7683Qq6xT653drQ:bx0v9op) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L46)):*
> ```Markdown
> Future is an abstraction of an asynchronous task. Each future has to be defined using
> `FUTURE(_name, _data_type, _output_type)` macro. It takes future name `_name`,
> input data type `_data_type` and output data type `_output_type`.
> ```

You could add a link to the manpage somewhere here.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 67 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0VksH65897jAZ-9ZxM:-N0VksH78Ty6Q0R64q_G:b14mc9d) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L67)):*
> ```Markdown
> `state` indicates whether the lookup function should search for an existing or unoccupied hashmap entry.
> 
> The output data `struct hashmap_lookup_output` contains a pointer to the hashmap entry `hm` that points
> ```

`hm` -> `hme`
___
*[`content/blog/2022/masync-ex-hashmap.md` line 84 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vl_P-ALnn06C4JSLE:-N0Vl_P-ALnn06C4JSLF:b-wy80d) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L84)):*
> ```Markdown
> of a particular future.
> 
> >Notifier feature is not used in hashmap example, hence it will be not discussed. For more information
> ```

in the hashmap example
___
*[`content/blog/2022/masync-ex-hashmap.md` line 84 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vlh0s-O-N_Ba5vzbG:-N0Vlh0s-O-N_Ba5vzbH:b-aqvj8k) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L84)):*
> ```Markdown
> of a particular future.
> 
> >Notifier feature is not used in hashmap example, hence it will be not discussed. For more information
> ```

`it will not be`
___
*[`content/blog/2022/masync-ex-hashmap.md` line 358 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vq-egAsw56qpOqSS_:-N0Vq-egAsw56qpOqSSa:bif67fp) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L358)):*
> ```Markdown
> ```
> After initializing each chained future entry, we initialize the chained future by passing
> its address to the `FUTURE_CHAIN_INIT(...)` macro. `hashmap_lookup_output` future is now
> ```

`hashmap_lookup_output` -> `hashmap_lookup_lock_entry_fut` (?)
___
*[`content/blog/2022/masync-ex-hashmap.md` line 395 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Vq_hc9gnB9pdXVWEC:-N0Vq_hc9gnB9pdXVWED:b-ttu9kb) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L395)):*
> ```Markdown
> }
> ```
> We can find an example of a lazy entry initialization in the `hashmap_get_copy_fut` future.
> ```

this first intro sentence could be written before the actual code of that future.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 407 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0VquMQ5YjxanqolUyM:-N0VquMQ5YjxanqolUyN:b-9ez7c0) ([raw file](https://github.com/pmem/pmem.github.io/blob/d5e3f07387fd2c0a64d8af1ca48eebd60e3d9b13/content/blog/2022/masync-ex-hashmap.md#L407)):*
> ```Markdown
> Now, we pass along an address of the `set_state_init_for_get(...)` function that is responsible
> for initializaing the `set_state` entry. Future initialization function has to conform to the
> `typedef void (*future_init_fn)(void *future, struct future_context *chain_fut, void *arg)`
> ```

you already wrote that part about conforming (above)

**Review by lukaszstolarczuk — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237)* status: 0 of 1 files reviewed, 7 unresolved discussions (waiting on @kswiecicki, @lukaszstolarczuk, and @pbalcer)

*[a discussion](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Zh54aBimkfAkG_XyO:-N0Zh54aBimkfAkG_XyP:btxum0y) (no related file):*
you could push this blog post to your fork and generate a preview ( https://kswiecicki.github.io/blog ) to see if everything is rendered correctly.

---
___
*[`content/blog/2022/masync-ex-hashmap.md` line 470 at r3](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZgusZ01XVbx-DkFhm:-N0ZgusZ01XVbx-DkFhn:b32r3n0) ([raw file](https://github.com/pmem/pmem.github.io/blob/64d1083b36d7a50fdb4a8a0cb5901de37e338493/content/blog/2022/masync-ex-hashmap.md#L470)):*
> ```Markdown
> 
> In summary, we have covered the future creation and initialization. We have also discussed the
> concept of chained futures and how we should initialize its entries. Lastly, we have talked
> ```

`chained futures ... its entries` either `future` or `their`
___
*[`content/blog/2022/masync-ex-hashmap.md` line 479 at r3](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZgqXT8voPpit78QLu:-N0ZgqXUESnbLHCGTdFl:b2kv30d) ([raw file](https://github.com/pmem/pmem.github.io/blob/64d1083b36d7a50fdb4a8a0cb5901de37e338493/content/blog/2022/masync-ex-hashmap.md#L479)):*
> ```Markdown
> [masync-hashmap]: https://github.com/kswiecicki/libuasync/blob/masync-example-hashmap/examples/hashmap/hashmap.c
> [masync-future-manpage]: https://github.com/pmem/miniasync/blob/master/doc/miniasync_future.7.md
> [murmurmash3-wiki]: https://en.wikipedia.org/wiki/MurmurHash
> ```

missing EOF

**Comment by kswiecicki** [comment]
*[a discussion](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Zh54aBimkfAkG_XyO:-N0ZnNkp327_muot-v6v:b4t5j7r) (no related file):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

you could push this blog post to your fork and generate a preview ( https://kswiecicki.github.io/blog ) to see if everything is rendered correctly.
</blockquote></details>

I was having problems with `github-pages` deployment, so I was testing this blogpost locally.
Anyway, I see that the deployment started working. The blog post is available here: https://kswiecicki.github.io/blog/2022/04/basic-asynchronous-hashmap-with-miniasync-library/

---

**Review by kswiecicki — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237)* status: 0 of 1 files reviewed, 7 unresolved discussions (waiting on @lukaszstolarczuk and @pbalcer)

___
*[`content/blog/2022/masync-ex-hashmap.md` line 470 at r3](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZgusZ01XVbx-DkFhm:-N0Zo-8p-h9wJMnmLTH2:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/64d1083b36d7a50fdb4a8a0cb5901de37e338493/content/blog/2022/masync-ex-hashmap.md#L470)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

`chained futures ... its entries` either `future` or `their`
</blockquote></details>

Done.
___
*[`content/blog/2022/masync-ex-hashmap.md` line 479 at r3](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZgqXT8voPpit78QLu:-N0Zo0CLEAIHXWeVOgV2:b-896fix) ([raw file](https://github.com/pmem/pmem.github.io/blob/64d1083b36d7a50fdb4a8a0cb5901de37e338493/content/blog/2022/masync-ex-hashmap.md#L479)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

missing EOF
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
:+1: 

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237#-:-N0ZrRuvDrmoAPTlDxQZ:b-8fx2u7)* status: all files reviewed, 5 unresolved discussions (waiting on @kswiecicki and @pbalcer)

*[a discussion](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0Zh54aBimkfAkG_XyO:-N0Zr6Dy-XFiQ-YPJa5y:b-fmk5xy) (no related file):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

I was having problems with `github-pages` deployment, so I was testing this blogpost locally.
Anyway, I see that the deployment started working. The blog post is available here: https://kswiecicki.github.io/blog/2022/04/basic-asynchronous-hashmap-with-miniasync-library/
</blockquote></details>

thx, it looks cool!

---
___
*[`content/blog/2022/masync-ex-hashmap.md` line 477 at r4](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZrCdK795hrCpIn80i:-N0ZrCdK795hrCpIn80j:blruvil) ([raw file](https://github.com/pmem/pmem.github.io/blob/0ddadbf70863b40eb36ea96747b6d5e658add6e7/content/blog/2022/masync-ex-hashmap.md#L477)):*
> ```Markdown
> [masync]: https://github.com/pmem/miniasync
> [masync-examples]: https://github.com/pmem/miniasync/tree/master/examples
> [masync-hashmap]: https://github.com/kswiecicki/libuasync/blob/masync-example-hashmap/examples/hashmap/hashmap.c
> ```

heh, just don't forget to bump this to 'pmem' link (when merged)

**Comment by kswiecicki** [comment]
*[`content/blog/2022/masync-ex-hashmap.md` line 477 at r4](https://reviewable.io/reviews/pmem/pmem.github.io/237#-N0ZrCdK795hrCpIn80i:-N0Zw9Wn0FUAs5YDZH4R:bwf0ggl) ([raw file](https://github.com/pmem/pmem.github.io/blob/0ddadbf70863b40eb36ea96747b6d5e658add6e7/content/blog/2022/masync-ex-hashmap.md#L477)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

heh, just don't forget to bump this to 'pmem' link (when merged)
</blockquote></details>

Sure :)

**Review by pbalcer — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237#-:-N0Zx5881v_tML5gfHUg:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @lukaszstolarczuk)

**Review by pbalcer — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237)* status: all files reviewed, 1 unresolved discussion (waiting on @lukaszstolarczuk)

File: content/blog/2022/masync-ex-hashmap.md [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+};
+struct hashmap_lookup_lock_entry_output {
+	struct hashmap_entry *hme;
+};
+FUTURE(hashmap_lookup_lock_entry_fut, struct hashmap_lookup_lock_entry_data,
+		struct hashmap_lookup_lock_entry_output);
+```
+Here, `struct hashmap_lookup_lock_entry_fut` future consists of three other futures,
+`struct hashmap_lookup_fut` that was discussed, `struct hashmap_entry_set_state_fut`,
+and 'struct chain_entries_rerun_fut' that follow the same principles the `lookup` future does.
```
Comment (DamianDuy) [inline content/blog/2022/masync-ex-hashmap.md:284]: I think `struct chain_entries_rerun_fut` was probably supposed to be inside the `  signs.

**Review by DamianDuy — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/237#-:-N2Gqd6Z8StPn4OtgBw1:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @kswiecicki and @lukaszstolarczuk)

File: content/blog/2022/masync-ex-hashmap.md [hunk 2]
```diff
@@ -0,0 +1,476 @@
+---
+# Blog post title
+title: "Basic asynchronous hashmap with Miniasync library"
+
+# Blog post creation date
+date: 2022-04-25T10:00:00+00:00
```
Comment (pbalcer) [inline content/blog/2022/masync-ex-hashmap.md:6]: please update so that blogs are sorted chronologically on our website.

**Pull request merged**